### PR TITLE
wolfssl: add config flag for Curve448

### LINKED ISSUE #18812: strongswan: missing Curve448 with wolfssl
Maintainer: @pprindeville @Thermi
Environment: OpenWrt 22.03-rc4, Linksys E8450 (aarch64 mt7622)

Description: Using WolfSSL for Strongswan does not enable the Curve448 option. Using OpenSSL does.

```
# swanctl --list-algs
encryption:
  AES_CBC[aes]
  AES_ECB[aes]
  3DES_CBC[des]
  DES_CBC[des]
  DES_ECB[des]
  RC2_CBC[rc2]
  AES_CTR[wolfssl]
  NULL[wolfssl]
integrity:
  HMAC_MD5_96[wolfssl]
  HMAC_MD5_128[wolfssl]
  HMAC_SHA1_96[wolfssl]
  HMAC_SHA1_128[wolfssl]
  HMAC_SHA1_160[wolfssl]
  HMAC_SHA2_256_128[wolfssl]
  HMAC_SHA2_256_256[wolfssl]
  HMAC_SHA2_384_192[wolfssl]
  HMAC_SHA2_384_384[wolfssl]
  HMAC_SHA2_512_256[wolfssl]
  HMAC_SHA2_512_512[wolfssl]
  AES_XCBC_96[xcbc]
aead:
  AES_GCM_16[wolfssl]
  AES_GCM_12[wolfssl]
  AES_CCM_16[wolfssl]
  AES_CCM_12[wolfssl]
  AES_CCM_8[wolfssl]
  CHACHA20_POLY1305[wolfssl]
hasher:
  HASH_SHA1[sha1]
  HASH_SHA2_224[sha2]
  HASH_SHA2_256[sha2]
  HASH_SHA2_384[sha2]
  HASH_SHA2_512[sha2]
  HASH_MD5[md5]
  HASH_SHA3_224[wolfssl]
  HASH_SHA3_256[wolfssl]
  HASH_SHA3_384[wolfssl]
  HASH_SHA3_512[wolfssl]
prf:
  PRF_KEYED_SHA1[sha1]
  PRF_HMAC_MD5[wolfssl]
  PRF_HMAC_SHA1[wolfssl]
  PRF_HMAC_SHA2_256[wolfssl]
  PRF_HMAC_SHA2_384[wolfssl]
  PRF_HMAC_SHA2_512[wolfssl]
  PRF_FIPS_SHA1_160[fips-prf]
  PRF_AES128_XCBC[xcbc]
xof:
drbg:
dh:
  ECP_256[wolfssl]
  ECP_384[wolfssl]
  ECP_521[wolfssl]
  ECP_224[wolfssl]
  MODP_3072[wolfssl]
  MODP_4096[wolfssl]
  MODP_2048[wolfssl]
  MODP_2048_224[wolfssl]
  MODP_2048_256[wolfssl]
  MODP_1536[wolfssl]
  MODP_1024[wolfssl]
  MODP_1024_160[wolfssl]
  MODP_768[wolfssl]
  MODP_CUSTOM[wolfssl]
  CURVE_25519[wolfssl]
  MODP_6144[gmp]
  MODP_8192[gmp]
rng:
  RNG_WEAK[wolfssl]
  RNG_STRONG[random]
  RNG_TRUE[random]
nonce-gen:
  NONCE_GEN[nonce]
```

According to the [Strongswan docs](https://docs.strongswan.org/docs/5.9/config/IKEv2CipherSuites.html#_modern_elliptic_curve_groups), Curve448 should work with WolfSSL. WolfSSL in OpenWrt 22.03 is 5.9.0, which does support it (since wolfssl 4.8.0: https://github.com/wolfSSL/wolfssl/blob/master/ChangeLog.md#wolfssl-release-480-july-09-2021)

My suspicion is that we aren't passing in a feature enable flag to WolfSSL's configure script like we are Ed25519. See https://github.com/openwrt/openwrt/blob/eac1b8ab699eb3383e6d5d39b53fa9f32ee6eb60/package/libs/wolfssl/Makefile#L121

### PR DESCRIPTION
**Pull request by lowjoel**
This enables building WolfSSL with Curve448, which can be used by Strongswan. This has been tested on a Linksys E8450, running OpenWrt 22.03-rc4.

This allows parity with OpenSSL, which already supports Curve448 in OpenWrt 21.02.

Fixes openwrt/packages#18812.

Signed-off-by: Joel Low <joel@joelsplace.sg>

### CODE REVIEW

**Comment by hauke** [comment]
How much bigger is the libwolfssl application after adding this change?

**Comment by lowjoel** [comment]
@hauke On aarch64, exactly 16KiB. Not sure if this is precise enough:

22.03-rc4:
```
-rwxr-xr-x    1 root     root       1180364 Jun 10 15:20 /usr/lib/libwolfssl.so.5.3.0.ee39414e*
```

local compile:
```
-rwxr-xr-x 1 joel joel 1196748 Jun 24 18:12 libwolfssl.so.5.3.0.5f7518c0*
```

**Comment by Ansuel** [comment]
@lowjoel can you reorganize the commit description with 

```
[ commit description]

[ Fixes tag ] (change your Addresses  with Fixes)
[ Sob ]
```

This way the issue is linked and it will get autoclosed

**Comment by lowjoel** [comment]
Sorry @Ansuel - I've fixed it; is this what you meant in your description?

Thanks!

**Comment by cotequeiroz** [comment]
I'm still debating with myself if this should be done or not.

Why should a basic package get 4k bigger for everyone else that does not need it?  That's my main question.  While not adding much (~4k, you say), I don't know if the feature will be popular enough to justify it.

Strongswan is already a large package, so I think wolfssl can be an option for a slim, basic setup; more advanced features can be used with openssl.  

I do not have objections with it myself.  If you can show me that it is indeed a popular, or necessary feature, then sure it should be added.

**Comment by cotequeiroz** [comment]
> FWIW, TLS 1.3 includes it in its standard list of algorithms. https://datatracker.ietf.org/doc/html/rfc8446#page-41.

The section you pointed to is for ed448 (signature), which is not being enabled here, but you probably meant 4.2.7 which is for x448 (key exchange).  Neither algorithm is designated as MUST (as P-256 is) or SHOULD (as x25519) be supported (section 9.1).  Since they provide stronger security than ed/x25519, there should come a time when they have to be enabled.  I just don't think the time is now.  

The original motivation for this PR was to use it for IKEv2 key agreement in strongswan, and I can't gauge its popularity.  Now that you mentioned TLS 1.3, I can at least try.  I took the list of browser capabilities from https://api.ssllabs.com/api/v3/getClients, and the only entry in the list of 158 (most of them outdated) agents supporting x448 is OpenSSL 1.1.1c.

What I would be in favor, is to add curve448 as an optional (`default no`) build option.

This is just my opinion.  Core openwrt members will have the last say on this.  

I've compared the actual image size for Archer C6 V2, and got an 8k difference in ipk size, corresponding to a 5kb difference in the final image.  I usually see comparisons using the ipk size, but I liked your approximation.

**Comment by lowjoel** [comment]
> The section you pointed to is for ed448 (signature), which is not being enabled here, but you probably meant 4.2.7 which is for x448 (key exchange).

Woops! Thanks for checking.

> Neither algorithm is designated as MUST (as P-256 is) or SHOULD (as x25519) be supported (section 9.1). Since they provide stronger security than ed/x25519, there should come a time when they have to be enabled. I just don't think the time is now.

That's fair.

> The original motivation for this PR was to use it for IKEv2 key agreement in strongswan, and I can't gauge its popularity. Now that you mentioned TLS 1.3, I can at least try. I took the list of browser capabilities from https://api.ssllabs.com/api/v3/getClients, and the only entry in the list of 158 (most of them outdated) agents supporting x448 is OpenSSL 1.1.1c.

Yeah. For reference though, OpenSSL 1.1.1c is from May 2019.

> What I would be in favor, is to add curve448 as an optional (`default no`) build option.

If that's the direction that the core members decide I'll be happy to change, but I'd rather avoid flip flopping. Do we have to mention anyone, or do I just wait? Not sure what the convention is here

> I've compared the actual image size for Archer C6 V2, and got an 8k difference in ipk size, corresponding to a 5kb difference in the final image. I usually see comparisons using the ipk size, but I liked your approximation.

😂 thanks!

**Comment by Ansuel** [comment]
> If that's the direction that the core members decide I'll be happy to change, but I'd rather avoid flip flopping. Do we have to mention anyone, or do I just wait? Not sure what the convention is here

Honestly i agree... if someone will run a vpn on their router than he for sure have plenty of space so installing full openssl is not a problem. For me it's ok to add option but i'm hesitant of enabling it by default considering there are alternative.

If it's not a problem i would ask you to add the option but keep it disabled by default.

**Comment by lowjoel** [comment]
@Ansuel option added but defaults to `n`.

I'm not sure that this still "fixes packages#18812" though, from the perspective of a person setting up Strongswan it's going to become "compile wolfssl with the flag set to y". Happy to change to a different keyword if you have a suggestion.

File: package/libs/wolfssl/Config.in [hunk 1]
```diff
@@ -51,6 +51,10 @@ config WOLFSSL_HAS_ECC25519
 	bool "Include ECC Curve 25519 support"
 	default y
 
+config WOLFSSL_HAS_ECC448
+	bool "Include ECC Curve 448 support"
+	default n
```
Comment (hauke) [inline package/libs/wolfssl/Config.in:56]: `default n` is the default you can just leave the line out.

**Comment by hauke** [comment]
Thank you for your patch, I applied it to master in 925c73aab3c1514da9f4cec97badb91113631dbf.

**Pull request closed without merging**